cmd/shimdiag/shimdiag.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/containerd/ttrpc"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/windows"
+)
+
+const (
+	shimPrefix = `\\.\pipe\ProtectedPrefix\Administrators\containerd-shim-`
+	shimSuffix = `-pipe`
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "shimdiag"
+	app.Usage = "runhcs shim diagnostic tool"
+	app.Commands = []cli.Command{
+		listCommand,
+		execCommand,
+	}
+	if err := app.Run(os.Args); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func findPipes(pattern string) ([]string, error) {
+	path := `\\.\pipe\*`
+	path16, err := windows.UTF16FromString(path)
+	if err != nil {
+		return nil, err
+	}
+	var data windows.Win32finddata
+	h, err := windows.FindFirstFile(&path16[0], &data)
+	if err != nil {
+		return nil, &os.PathError{Op: "FindFirstFile", Path: path, Err: err}
+	}
+	var names []string
+	for {
+		name := `\\.\pipe\` + windows.UTF16ToString(data.FileName[:])
+		if matched, _ := filepath.Match(pattern, name); matched {
+			names = append(names, name)
+		}
+		err = windows.FindNextFile(h, &data)
+		if err == windows.ERROR_NO_MORE_FILES {
+			break
+		}
+		if err != nil {
+			return nil, &os.PathError{Op: "FindNextFile", Path: path, Err: err}
+		}
+	}
+	return names, nil
+}
+
+func findShims(name string) ([]string, error) {
+	pipes, err := findPipes(shimPrefix + name + "*" + shimSuffix)
+	if err != nil {
+		return nil, err
+	}
+	for i, p := range pipes {
+		pipes[i] = p[len(shimPrefix) : len(p)-len(shimSuffix)]
+	}
+	sort.Strings(pipes)
+	return pipes, nil
+}
+
+func findShim(name string) (string, error) {
+	if strings.ContainsAny(name, "*?\\/") {
+		return "", fmt.Errorf("invalid shim name %s", name)
+	}
+	shims, err := findShims(name)
+	if err != nil {
+		return "", err
+	}
+	if len(shims) == 0 {
+		return "", fmt.Errorf("no such shim %s", name)
+	}
+	if len(shims) > 1 && shims[0] != name {
+		return "", fmt.Errorf("multiple shims beginning with %s", name)
```

Comment:
It seems this would fail if we ever had a shim naming scheme that made it possible for some shim names to be prefixes of others. Would it be better to just search the returned shims for an exact match here, in case we ever adopt such a scheme?

Reply:
That's why I have the `shims[0] != name` clause--if you pass the exact name, you get what you asked for.

It's a real pain to copy and paste the whole shim name in every time. I learned this from Docker and crictl.

Reply:
My point was that we're taking a dependency on there not being valid shim names that are prefixes of other shim names. For instance, if you have shim `a` and shim `ab`, this function would not work if `ab` is returned before `a`.

However, as I was writing this, I realized the shim list is sorted, so I don't think this case could ever occur. :)